#89 Track parser

nolmoonen opened the pull request:
#### Deployment instructions
None.

#### What have I done
Completed the track parser, which uses the path service and a list of track entries to (mainly )create an list of segments ordered by appearance. This is the content of the Route service, and the parser initializes that service.

#### Testing
Run unit tests.

Extremelyd1 commented on src/app/backend/data-representation/route.ts:
```diff
@@ -1,33 +1,142 @@
+        let i = 0;
+        for (const offset of this.segmentsOffset) {
+            if (offset > position) {
+                return this.segments[i - 1];
```
This does not work when the input for this function is a position before any segment. Then the array is index at `-1`. When fixing this, keep in mind that there should also be a test case catching this.

nolmoonen replied:
Fixed by throwing error if `position < 0`, added test case as well.

Extremelyd1 commented on src/app/backend/data-representation/route.ts:
```diff
@@ -1,33 +1,142 @@
+     * @throws IllegalArgumentError if size mismatch
+     * @param offset to set.
+     */
+    setSegmentsOffset(offset: number[]): void {
```
Don't forget to rename this, when addressing the comment about the variable name above.

nolmoonen replied:
Renamed comments and local variables as well!

Extremelyd1 commented on src/app/backend/data-representation/route.ts:
```diff
@@ -1,33 +1,142 @@
+            if (segment === thisSegment) {
+                return this.segmentsOffset[i];
+            }
+            i++;
```
There is a way to find the index of a object in an array, namely `indexOf()`. This would reduce the code here and make it more readable.

nolmoonen replied:
Replaced by a for `i` loop, which is also more intuitive.

Extremelyd1 commented on src/app/backend/file-parser/track-parser.service.ts:
```diff
@@ -14,36 +18,239 @@ export class TrackParserService {
+
+            if (trackEntryElement.children[0].tagName !== 'Name') {
+                throw new ParserError(
+                    'TrackEntry entry number ' + i + ' must have \"Name\" as 1st tag.'
```
No need to escape the `"` character here.

nolmoonen replied:
Removed it.

nolmoonen commented on src/app/backend/data-representation/route.ts:
```diff
@@ -1,33 +1,142 @@
+    /**
+     * List containing the global absolute start position of every segment by index.
+     */
+    private segmentsOffset: number[];
```
Renamed to `segmentStartPositions`.

nolmoonen commented on src/app/backend/file-parser/track-parser.service.ts:
```diff
@@ -14,36 +18,239 @@ export class TrackParserService {
         const trackEntries: TrackEntry[] = [];
 
-        // loop over all track entry elements
         const numberTrackEntries = trackData[0].children.length;
```
This is because the file structure: one element `TrackData`, containing multiple `TrackEntry` elements.

nolmoonen commented on src/app/backend/file-parser/track-parser.service.ts:
```diff
@@ -14,36 +18,239 @@ export class TrackParserService {
+    /**
+     * Calculate and set the segment offset in the route service.
+     */
+    calculateAndSetSegmentsOffsetAndLength(): void {
```
Is tested implicitly, but added unit test as well.

nolmoonen commented on src/app/backend/file-parser/track-parser.service.ts:
```diff
@@ -14,36 +18,239 @@ export class TrackParserService {
+     * @throws ParserError see {@code fromFile}
+     * @throws IllegalArgumentError see {@code fromFile}
+     */
+    parse(trackText: string): void {
```
Do you have suggestions on how to break it up? I am not sure how to do this.

Dacaspex replied:
This function is indeed quite large and perhaps should be split up in more, smaller, functions

nolmoonen replied:
Addressed this in latest commit, please see if you agree.

nolmoonen commented on src/app/backend/file-parser/path-parser.ts:
```diff
@@ -10,12 +10,13 @@ import { Inject, Injectable } from '@angular/core';
+        const segmentCPs = new DOMParser().parseFromString(txt, 'text/xml').getElementsByTagName('SegmentAppearance');
         // DOMParse for Points Of Interest
         const segmentPOIs = new DOMParser().parseFromString(txt, 'text/xml').getElementsByTagName('Segment');
 
```
Will ignore this comment like you said.

nolmoonen commented on src/app/backend/file-parser/path-parser.ts:
```diff
@@ -10,12 +10,13 @@ import { Inject, Injectable } from '@angular/core';
 })
```
Not mine, fixed anyways.

nolmoonen commented on src/app/backend/file-parser/track-parser.service.ts:
```diff
@@ -14,36 +18,239 @@ export class TrackParserService {
+        for (const segment of this.route.getSegments()) {
+            // next offset is current offset plus length of current segment
+            offset.push(lastOffset);
+            lastOffset = lastOffset + segment.getLength();
```
Sure.

nolmoonen commented on src/app/backend/file-parser/path-parser.ts:
```diff
@@ -51,6 +55,10 @@ export class PathParser {
             const segId = Number(seg.children[0].firstChild.nodeValue);
+
+            // highestId = max(highestId, segId)
+            highestId = segId > highestId ? segId : highestId;
```
Only because you say so.

nolmoonen commented on src/app/backend/data-representation/route.ts:
```diff
@@ -1,33 +1,142 @@
+        }
+
+        let i = 0;
+        for (const thisSegment of this.segments) {
```
Have changed this already in response to comment of @Extremelyd1 .

nolmoonen commented on src/app/backend/data-representation/path.ts:
```diff
@@ -30,4 +39,20 @@ export class Path {
+    }
+
+    /**
+     * Gets the highest id.
```
Same^.

nolmoonen commented on src/app/backend/data-representation/path.ts:
```diff
@@ -30,4 +39,20 @@ export class Path {
     }
+
+    /**
+     * Sets the highest id.
```
Added brief description to variable declaration.

nolmoonen commented:
Also removed the weird way of copying and reversing an array, it is a bit better now.

nolmoonen commented on src/app/backend/file-parser/track-parser.service.ts:
```diff
@@ -14,36 +18,239 @@ export class TrackParserService {
+        }
+
+        // reverse the order so the positions are increasing
+        reversedPOIPositions.reverse();
```
Changed this.

nolmoonen commented on src/app/backend/file-parser/track-parser.service.ts:
```diff
@@ -14,36 +18,239 @@ export class TrackParserService {
+            reversedControlPoints.push(controlPoint);
+        }
+        // reverse the order so the positions are increasing
+        reversedControlPoints.reverse();
```
Same^.

nolmoonen commented on src/app/backend/file-parser/track-parser.service.ts:
```diff
@@ -14,36 +18,239 @@ export class TrackParserService {
+        );
+
+        // update highest id
+        this.path.setHighestId(this.path.getHighestId() + 1);
```
Good suggestion! Changed it.

nolmoonen commented on src/app/backend/file-parser/track-parser.service.ts:
```diff
@@ -14,36 +18,239 @@ export class TrackParserService {
+    }
+
     /**
      * Reads an array of TrackEntries in from a string containing a track data file, and an array of Points of Interest.
```
Yes, updated only the parameter tags, but not the description.

nolmoonen commented on src/app/backend/data-representation/path.ts:
```diff
@@ -9,16 +9,25 @@ export class Path {
 
     // Getters & Setters
-    setPath(segments: Segment[], pois: POI[]) {
+    setPath(segments: Segment[], pois: POI[]): void {
```
Changed it so it is set when the path is set, and updated by an incrementer.

nolmoonen commented on src/app/backend/data-representation/path.spec.ts:
```diff
@@ -21,4 +21,10 @@ describe('Path', () => {
+        path.setHighestId(42);
+        expect(path.getHighestId()).toBe(42);
+    });
 });
```
What do you mean by invariant? Is this something to be checked programmatically?

nolmoonen commented on src/app/backend/data-representation/route.ts:
```diff
@@ -1,33 +1,142 @@
+            if (position <= this.segments[0].getLength()) {
+                return this.segments[0];
+            } else {
+                throw new IllegalStateError('position is higher than sum of segments');
```
Fixed.

nolmoonen commented on src/app/backend/data-representation/route.ts:
```diff
@@ -1,33 +1,142 @@
+    }
+
+    /**
+     * Returns a list of segments.
```
Fixed.

Dacaspex commented on src/app/backend/data-representation/route.ts:
```diff
@@ -1,33 +1,133 @@
+        if segment is last one in array
+         */
+        const lastStartPosition = this.segmentStartPositions[this.segmentStartPositions.length - 1];
+        const lastSegement = this.segments[this.segments.length - 1];
```
[Typo](https://www.youtube.com/watch?v=EShUeudtaFg)

nolmoonen replied:
Changed.

Dacaspex commented on src/app/backend/data-representation/route.ts:
```diff
@@ -1,33 +1,133 @@
+         */
+        const lastStartPosition = this.segmentStartPositions[this.segmentStartPositions.length - 1];
+        const lastSegement = this.segments[this.segments.length - 1];
+        if (lastStartPosition + lastSegement.getLength() > position) {//
```
`//` can be removed

nolmoonen replied:
Did this.

Dacaspex commented on src/app/backend/file-parser/track-parser.service.ts:
```diff
@@ -15,35 +20,236 @@ export class TrackParserService {
-    fromFile(trackText: string, arrayOfPOI: POI[]): TrackEntry[] {
+    parse(trackText: string): void {
+        // filtered out string
+        const FILTERED_STRING = 'InterfacePoint';
```
Why are we filtering out track entries that contain `InterfacePoint`?

nolmoonen replied:
Explained to you verbally, but added a more detailed comment.

Dacaspex commented on src/app/backend/file-parser/track-parser.service.ts:
```diff
@@ -15,35 +20,236 @@ export class TrackParserService {
+         Filters out not needed track entries: removes all that include the FILTERED_STRING, removes all that occur more
+         than one time. Only used by parse function.
+         */
+        function getFilteredTrackEntries(returnedTrackEntries: TrackEntry[]): TrackEntry[] {
```
Function inside a function, why not a separate function?  I would avoid this

nolmoonen replied:
Discussed this with you, made the function private.

nolmoonen commented on src/app/backend/file-parser/track-parser.service.ts:
```diff
@@ -14,36 +18,239 @@ export class TrackParserService {
-        // loop over all track entry elements
         const numberTrackEntries = trackData[0].children.length;
+
+        // ERROR CHECK - Ensure the structure of each TrackEntry entry is correct
```
Changed to use Pieters function.

nolmoonen commented:
Refactored `track-parser.parse`: broke function up in three private functions, added extensive contracts/comments. Also refactored `route.getSegment` slightly.
